### DISCUSSION TITLE
How to create a standalone application to control Stellarium

### DISCUSSION THREAD

**Discussion by PotatoTroll** [Stellarium/stellarium#3752] [General]
I would like to create a standalone application that would control the Stellarium similar to Stellarium Console: move Stellarium to second monitor, focus on selected planets, toggle Stellarium options like atmosphere and fog on and off, chage time and so on. This would be an interesting project to work on and I would make it open source if I will manage to create it.

However the issue is that none of the documentation is clear enough on how could I achieve that, Could someone help me on achieving this or tell what is the best way to approach this?

**Comment by gzotti** [comment]
What is Stellarium Console? OK, yes, I have seen their website before. You will not find any information about this program or interface or internals from our side. 

The RemoteControl plugin allows pretty much control of the program via HTTP interface from your web browser. The interface can of course be extended/improved, adapted to various screen sizes (FullHD monitor, 10" tablet, 7" tablet, 6" smartphone screen, ..., or some adaptive do-it-all solution?) Requires HTML/CSS/Javascript skills, and may require thorough study of the available classes and scriptable module properties (our API and Doxygen).

**Reply by PotatoTroll** [reply]
Ok that makes sense.
Of course the RC interface as you say could be extended to have a Stelarium Console like functionality, but I doubt that's allowed(or even needed). Another option would be to create and maintain a fork of stellarium just for that, but it would be easier just to develop a 3rd party app using something like electron that uses just the RC api and a more modern web dev stack(although just opening the RC interface in browser from anywhere does make it better). Does it make any sense?
Also wondering if something like this would be useful to anyone else, Stelarium Console is used in a local planetarium here to  teach by demoing planets/constelations and stuff but the biggest downside (other that it being a paid software) is that it uses a really old stellarium version.

**Reply by alex-w** [reply]
It may be very helpful for lecturers in real planetariums

**Reply by gzotti** [reply]
A third-party app will always have to be maintained by third-party developers. A well-designed new interface that extends the current browser-based RC plugin user interface can be maintained by us. Nowadays, a few new functions appear, or some functions are still occasionally renamed, these are small enough changes for maintainers. OTOH, a big change is coming with skycultures in the next year or so. 

Never having seen SC in practice, I don't know what's preventing you from implementing such UI inside the RC plugin, or what should be disallowed. (Are there patents registered about controlling a FOSS app?)  What are the requirements or characteristics of a "more modern web dev stack"? The UI files are hand-written HTML/JS and yes, are somewhat cumbersome to maintain/change.  If there are more modern tools that help designing/layouting/testing, we are happy to get pointed at them. You can also probably experiment with a fully new UI layout ("Responsive" web design) that overcomes recreation of our widget-based UIs with their screen space requirements and limitations. Maybe use more sliders than spinners. While development should be done in a fork of yours, in the end it shall be merged to the main app to receive further updates and see widest possible distribution. However, reading about Electron, I am also reading about software bloat, heavy resource use and security issues. Try to remain lightweight, think of Raspberry Pi4. (OK, not the usual target platform in a planetarium. But it would have 2 screens after all. Waiting for a tinker genius to show a sustainable self-modified fisheye projector attached to one RPi4/5 running Stellarium with RC touch control inside a Buckminster dome made of solar panels...)

The current RC API must remain, as it is already in use by a few 3rd-party programs. I think the design is good, with traceable state changes being forwarded to registered clients, so that it can be operated even by several clients. If something is missing, it can be extended of course. There is currently no RC support for oculars/telescope control, given that usually you would run desktop Stellarium on your field laptop, so we did not see the point here. Also some configurations like file locations or screenshot sizes/target directories are not implemented, just as taking screenshots is not something you would usually do while explaining the sky. (Unless there is need to distribute them after the show, OK. Can be added easily.)  If the webserver that we are using should have more capabilities, this module may be replaced. 

EDIT: There is also nothing in the browser UI similar to AstroCalc which has grown into a really big and interesting module in the last years. Here it is unclear to me whether everything should be replicated again in JS and data queries opened up in scriptable methods with their own data protocols. Does anyone need this functionality on a second screen? Note that it sometimes stalls the main view due to heavy computations, so for me its use in the planetarium while showing the sky to others is limited. On the other hand, a web interface where one could click onto data lines (events) in an ephemeris can be helpful. This is what bookmarks are doing now, so these should go into the RC functionality. And clicking into the eclipse panel to go to the best site is my second thought here... This separates a prepared show (using bookmarks) from spontaneous ideas (running program computing functionalities) one presenter may have.

Yes, this would be very helpful for small digital planetaria running Stellarium, so there is demand for a new UI for RC plugin better geared towards such installations. If possible, think of an operator/"tour guide to the night sky" not sitting on a big screen with 100 knobs and keyboard on the console, but sitting among the students with a 10" tablet as only interface device to the program, and maybe a laser pointer to explain/point at details. If the tablet itself acts as navigation aid by tablet motion, this may also be helpful (but guarded by a touch keypress! Sometimes physical buttons cannot be surpassed. Add some USB-connected customized 3D-printed hardware?).  Another UI could be running on a 20" touch screen in front of a regular 16:9 projection wall. And again more modes: operator/guide with all options, or an unsupervised kiosk-type installation for visitor operation with a preselection of options and automatic shows. Just a few ideas of what could be done if there is demand.

**Reply by PotatoTroll** [reply]
> I don't know what's preventing you from implementing such UI inside the RC plugin, or what should be disallowed.

Because I assumed someone takes a decision on everything going into Stellarium? It's not like I can change it in any way I want and expect all my changes to get merged in, right?

> What are the requirements or characteristics of a "more modern web dev stack"?

I meant something more related to a modern frontend tech stack based on React and other stuff that makes development faster. It can also be done with plain html/js, but that would be a pain.

Sorry, you went into some Stellarium details that I don't know much about, I'm in no way experienced with it, I only help the local planetarium so only familiar with the basics of it :)

**Reply by gzotti** [reply]
Yes, we (team) take decisions about what goes into Desktop Stellarium. Stellarium Control is not our project, though. You can write a fully new HTML/JS interface (or GUI written in any other language, Tk, Python, ...) that connects to the RC plugin's HTTP port. I recommend you are doing that new UI as new HTML/JS file in RC's directory, alternatively to index.html. Never having had anything to do with React or Electron, I don't know the implications of using those tools.  Do we then need added runtimes, does it work on all platforms, licenses, ....

**Reply by PotatoTroll** [reply]
I get that Stellarium Control is not your project, never implied that it is  :)

If we are talking about developing the a new GUI using React the only involvement is that it will probably require a new repo that will contain just the js code that later is compiled to plain html/js to be used in RC and respectively the build pipeline should be adjusted for that(or just manually copy the files to RC). But that could be to much for Stellarium for the existing functionality, unless you plan on really expanding the RC web app capabilities.

Anyway I get your point, I will try to look around and do something later, and if I will manage I'll get back here :)

**Reply by gzotti** [reply]
Ah OK. Then you would use a React-only repo for developing, and then we can either (1) *optionally* copy over/install the HTML/JS into the user's RC directory in a separate user download/install process (i.e., keep it outside the Stellarium project but e.g. recommend its usage in the User Guide) or (2) deploy this build as external dependency in our build process as default component as long as it works. If delivered with the regular install, I think the default visual style (color selection) should be close to our current style, but of course technically you can then use what React has to offer, and show what UI design in the 2020s has to offer.  

If the repo is not owned by us, I see some danger of eventual deprecation. I hope you are prepared for a decade or more of maintenance. Not many people use the RC interface (the HTML was built for an exhibition operator originally, but the API is now used by other programs), but I know a few planetarians who would like to see more/better interfacing for their installations.  I also have ideas, but not enough time. 

There is still the ancient Text User Interface. Maybe I can extend the RC plugin to incorporate the text display along the dome perimeter (and possibly other unique features), then the plugin could be phased out. Too much code for too few maintainers...

**Reply by PotatoTroll** [reply]
In terms of visual style, it would be helpful if you had someone who could sketch a design to follow :)
Also an interesting fact, Stellarium Console uses almost only red color, it looks ugly, but is probably used to help somehow by reducing the light inside the small planetariums or something, not sure..

**Reply by gzotti** [reply]
I only mean the color scheme, appearance (round corners etc) and icons of our own existing main GUI. This is encoded in our CSS files, and I think it should by default harmonize to appear as integral part. Alternative color schemes for daylight use, dark dome use etc.,  are possibly useful extensions, and radically redesigned UI elements could also find new friends.  Our GUI can be tweaked by loading a customized CSS, but it's for experts -- see User Guide.

Regarding menu layout, you can probably either replicate the current UI panels (so that all users know what's where; this strategy has been largely followed in RemoteControl's current web UI) or design something fresh that should be consistent, but may be entirely different as long as it is useful (speaking of Responsive web interface that looks good on all screen sizes including smartphones). The definition of useful may be a problem, though...

**Reply by gzotti** [reply]
@PotatoTroll any updates on your project?
